x/core/keeper/msg_server.go

```diff
@@ -0,0 +1,165 @@
+package keeper
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+
+	"cosmossdk.io/collections"
+	"cosmossdk.io/math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	vrf "github.com/sedaprotocol/vrf-go"
+
+	"github.com/sedaprotocol/seda-chain/x/core/types"
+)
+
+var _ types.MsgServer = msgServer{}
+
+type msgServer struct {
+	Keeper
+}
+
+func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+	return &msgServer{Keeper: keeper}
+}
+
+func (m msgServer) AddToAllowlist(goCtx context.Context, msg *types.MsgAddToAllowlist) (*types.MsgAddToAllowlistResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	if msg.Sender != m.GetAuthority() {
+		return nil, sdkerrors.ErrUnauthorized.Wrapf("unauthorized authority; expected %s, got %s", m.GetAuthority(), msg.Sender)
+	}
+
+	exists, err := m.isAllowlisted(ctx, msg.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+	if exists {
+		return nil, types.ErrAlreadyAllowlisted
+	}
+
+	err = m.addToAllowlist(ctx, msg.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO Add events
+
+	return &types.MsgAddToAllowlistResponse{}, nil
+}
+
+func (m msgServer) Stake(goCtx context.Context, msg *types.MsgStake) (*types.MsgStakeResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Verify stake proof.
+	var sequenceNum uint64
+	var isExistingStaker bool // for later use
+	staker, err := m.GetStaker(ctx, msg.PublicKey)
+	if err != nil {
+		if !errors.Is(err, collections.ErrNotFound) {
+			return nil, err
+		}
+	} else {
+		sequenceNum = staker.SequenceNum
+		isExistingStaker = true
+	}
+
+	hash, err := msg.MsgHash("", ctx.ChainID(), sequenceNum)
+	if err != nil {
+		return nil, err
+	}
+	publicKey, err := hex.DecodeString(msg.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+	proof, err := hex.DecodeString(msg.Proof)
+	if err != nil {
+		return nil, err
+	}
+	_, err = vrf.NewK256VRF().Verify(publicKey, proof, hash)
+	if err != nil {
+		return nil, types.ErrInvalidStakeProof.Wrapf("%s", err.Error())
+	}
```

Comment:
IMO might be best to put all these verification sections into functions. Makes the logic in the places it's done cleaner/if we need to update it we only need to touch it in one place.